libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs

```diff
@@ -902,170 +971,91 @@ private static (object value, string error) AccessIndex(object instance, int ind
             return (value, error);
         }
 
-        private static (object value, string error) ExtractElement(Expression expression, object state)
+        private static (object value, string error) ExtractElement(Expression expression, IMemoryScope state)
         {
             object value = null;
-            string error;
-            var instance = expression.Children[0];
-            var index = expression.Children[1];
-            object inst;
-            (inst, error) = instance.TryEvaluate(state);
-            if (error == null)
+
+            // if children 0 is Accessor, we will try to accumalate path
+            if (expression.Children[0].Type == ExpressionType.Accessor)
             {
-                object idxValue;
-                (idxValue, error) = index.TryEvaluate(state);
-                if (error == null)
+                try
                 {
-                    if (idxValue is int idx)
-                    {
-                        (value, error) = AccessIndex(inst, idx);
-                    }
-                    else if (idxValue is string idxStr)
+                    var (path, left) = TryAccumulatePath(expression, state);
+                    if (left == null)
                     {
-                        (value, error) = AccessProperty(inst, idxStr);
+                        // fully converted to path
+                        value = state.GetValue(path);
                     }
                     else
                     {
-                        error = $"Could not coerce {index}<{idxValue.GetType()}> to an int or string";
+                        // stop at somewhere
+                        var (newScope, error) = left.TryEvaluate(state);
+                        if (error != null)
+                        {
+                            return (null, error);
+                        }
+
+                        value = new SimpleObjectScope(newScope).GetValue(path);
                     }
                 }
-            }
-            return (value, error);
-        }
-
-        private static bool CanBeModified(object value, string property, int? expected)
-        {
-            var modifiable = false;
-            if (expected.HasValue)
-            {
-                // Modifiable list
-                modifiable = TryParseList(value, out var _);
-            }
-            else
-            {
-                // Modifiable object
-                modifiable = value is IDictionary<string, object>
-                    || value is IDictionary
-                    || value is JObject;
-                if (!modifiable)
+                catch (Exception e)
```

Comment:
Why was this necessary?  We are trying to keep exceptions out of the expression evaluation--what is likely to throw an exception?

Reply:
the exception here introduced is because the IMemoryScope interface is like

object GetValue(string path) // here may through exception

if we make this interface return (result, error), it feels like a little bit like go-back to C-style, and actually this interface here is totally called by internal evaluation process, so i wrap the exception into (result, error) again here, so I can keep the interface simpler. 

Do you think that's worth it, or we can make the IMemoryScope interface GetValue return (result, error) also?